File: 백준/Silver/15828. Router/Router.kt
package org.example

import java.util.LinkedList

fun main() {
    val br = System.`in`.bufferedReader()
    val n = br.readLine().toInt() // 버퍼 크기
    val router = LinkedList<Int>()
    var routerNowSize = 0 // 크기 추적용
    while (true) {
        val info = br.readLine().toInt()
        when (info) {
            -1 -> break
            0 -> {
                if (router.isNotEmpty()) { // 버퍼가 비어있을 때 0이 입력으로 안 들어온다고는 하지만 넣어봄
                    router.poll()
                    routerNowSize--
                }
            }

            else -> {
                if (routerNowSize == n) {
                    continue
                } else {
                    router.add(info)
                    routerNowSize++
                }
            }
        }
    }
    if (router.isNotEmpty()) {
        println(router.joinToString(" "))
    } else {
        println("empty")
    }

}

File: 백준/Silver/15828. Router/Router.py
import sys
from collections import deque

input = sys.stdin.readline

n = int(input())  # 버퍼 크기 N

router = deque()  # 음 크기 제한 어떻게 걸지. isfull 그런거 어케 하지
router_now_size = 0  # 라우터 크기 추적용. 매번 size 구하면 시간 걸리니깐
while True:
    info = int(input())
    if info == -1:
        break
    elif info == 0 and router:
        router.popleft()
        router_now_size -= 1
    else:
        if router_now_size == n:
            continue
        else:
            router.append(info)
            router_now_size += 1

if router:
    print(' '.join(map(str, router)))
else:
    print("empty")